src/source.ts

```diff
@@ -79,6 +83,29 @@ export default class Source extends React.Component<Props> {
     map.off('sourcedata', this.onData);
   };
 
+  public removeSource(): LayerWithBefore[] {
+    const { map } = this.context;
+
+    if (map.getSource(this.id)) {
+      let { layers = [] } = map.getStyle();
+
+      layers = layers
+        .map((layer, idx) => {
+          const { id: before } = layers[idx + 1] || { id: undefined };
+          return { ...layer, before };
+        })
+        .filter(layer => layer.source === this.id);
+
+      layers.forEach(layer => map.removeLayer(layer.id));
+
+      map.removeSource(this.id);
+
+      return layers.reverse();
```

Comment:
Why do you need to reverse the layers?

Reply:
This is because `map.addLayer` has an option for `before`, but not `after` ([`addLayer(layer, before?)`](https://www.mapbox.com/mapbox-gl-js/api#map#addlayer)). Say for example you have layer IDs that look like `[ 'asdf', 'zxcv', 'qwer' ]`, then we generate something like `[ { id: 'asdf', before: 'zxcv' }, { id: 'zxcv', before: 'qwer' }, { id: 'qwer', before: undefined } ]`.

If you try to add them back in the same order, `asdf` can't go before `zxcv` because it's not in there yet. So reversing it makes it so that `qwer` goes in at the end, then `zxcv` gets added in before `qwer`, etc.

I would love for the code to be more clear so this doesn't get changed later. Maybe we could add a comment, or can you think of a way to rewrite it that it's more obvious?